CORE-15285: Change sampling logic to favour flow traces

ZawarR3 opened the pull request:
The current configurability of traces is limited to a provided sample rate or unlimited sampling for all system requests, agnostic of their source or purpose. To reduce unnecessary data from being sampled, developers usually define a sample rate of their preference. However, this sample rate will be applied to all requests arbitrarily, meaning that the trace one is interested in might not get sampled.

When it comes to tracing as a whole, we are more interested in flow request sampling. With this change, we ensure that whatever sample rate is defined, is applied to all requests from URI endpoint `/flow` with method `POST,` as well as all other requests from the server. The change currently supports the following REST API versions: `v5_1,`, `v1`.

ZawarAliR3 committed:
CORE-15282: favour start flow in sampling

ZawarAliR3 committed:
Successfully implemented custom sampling
- ensured that sampleRate specified is applied to all requests AND /flow endpoints
- this ensures that atleast one flow request is sampled

ZawarAliR3 committed:
Added rule to support rest api version 1 and removed debug log messages

josephzunigadaly commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -90,6 +85,25 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String?, sam
         tracingBuilder.build().also(resourcesToClose::push)
     }
 
+    private fun sampler(samplesPerSecond: SampleRate): Sampler? = when (samplesPerSecond) {
```
Let's make the return type non-nullable

```suggestion
    private fun sampler(samplesPerSecond: SampleRate): Sampler = when (samplesPerSecond) {
```

josephzunigadaly commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -90,6 +85,25 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String?, sam
+        }
+
+        is Unlimited -> {
+            logger.info("Tracing will sample unlimited requests per second")
```
We will get these log messages 3 times now.

It's a minor thing, but I'd pull the log statements out into the `tracing` lambda so we only see them once.

ZawarR3 replied:
You want me to move the log statements into the by lazy code block for tracing initialisation?

ZawarR3 replied:
like this?: ```when (samplesPerSecond) {
            is PerSecond -> {
                logger.info("Tracing will sample ${samplesPerSecond.samplesPerSecond} requests per second")
            }
            is Unlimited -> {
                logger.info("Tracing will sample unlimited requests per second")
            }
        }```

josephzunigadaly replied:
yes please

ZawarAliR3 committed:
Made following changes after review:
- made Sampler() function return type not nullable
- moved logging statements from Sampler() function to tracing lambda so they do not get logged at every sampler instance

josephzunigadaly commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -90,6 +94,23 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String?, sam
+    }
+
+    private val serverSampler: SamplerFunction<HttpRequest> = HttpRuleSampler.newBuilder()
+        .putRule(and(methodEquals("POST"), pathStartsWith("/api/v5_1/flow")), sampler(samplesPerSecond))
+        .putRule(and(methodEquals("POST"), pathStartsWith("/api/v1/flow")), sampler(samplesPerSecond))
```
We need a better approach to identifying flow endpoints. This will stop matching in future releases.

I think we should merge this PR as it is and fix in a follow up PR.
